uvicorn/protocols/websockets/websockets_impl.py

```diff
@@ -236,6 +235,10 @@ async def asgi_receive(self):
             return {"type": "websocket.connect"}
 
         await self.handshake_completed_event.wait()
+        if self.closed_event.is_set():
+            # Handshake has failed.
+            return {"type": "websocket.disconnect", "code": self.initial_response[0]}
```

Comment:
That `code` is tricky… We actually need to consider two situations: A/ the handshake failed / got rejected, in that case `.initial_response` is available and we can use it like here, and B/ the connection was closed normally by the client sending a `websocket.close` event, and in that case `.initial_response` isn't available and we need to send the close code that was given by the client.